libvast/src/system/partition.cpp

```diff
@@ -54,42 +59,112 @@ partition::~partition() noexcept {
+             meta_data_.types.size(), "layouts and",
+             partition_type.fields.size(), "columns");
+  for (auto& f : partition_type.fields) {
+    qualified_record_field fqf{f.name, f.type};
+    indexers_.emplace(std::move(fqf), wrapped_indexer{});
+  }
   return caf::none;
 }
 
+void partition::finalize() {
+  // this->out().unregister(this);
+  // auto& out = this->out();
+  // auto silent = false; // Send shutdown signal on the stream
+  // for (auto& ind : indexers_)
+  //  out.remove_path(ind.second.slot, caf::none, silent);
+  for (auto& wi : indexers_) {
+    wi.second.buf.clear();
+  }
+}
+
 caf::error partition::flush_to_disk() {
   if (meta_data_.dirty) {
     // Write all layouts to disk.
-    if (auto err = save(nullptr, meta_file(), meta_data_))
+    if (auto err = save(nullptr, meta_file(), meta_data_, combined_type()))
       return err;
     meta_data_.dirty = false;
   }
-  // Write state for each layout to disk.
-  for (auto& kvp : table_indexers_)
-    if (auto err = kvp.second.flush_to_disk())
-      return err;
   return caf::none;
 }
 
+caf::actor& partition::indexer_at(size_t position) {
+  VAST_ASSERT(position < indexers_.size());
+  auto& [fqf, ip] = as_vector(indexers_)[position];
+  if (!ip.indexer) {
+    ip.indexer = state().make_indexer(column_file(fqf), fqf.type, id(),
+                                      &measurements_[position]);
+    VAST_ASSERT(ip.indexer != nullptr);
+  }
+  return ip.indexer;
+}
+
 // -- properties ---------------------------------------------------------------
 
-namespace {
+void partition::add(table_slice_ptr slice) {
+  VAST_ASSERT(slice != nullptr);
+  VAST_TRACE(VAST_ARG(slice));
+  meta_data_.dirty = true;
+  auto first = slice->offset();
+  auto last = slice->offset() + slice->rows();
+  auto& layout = slice->layout();
+  bool is_new = meta_data_.types.emplace(to_digest(layout), layout).second;
+  auto it = meta_data_.type_ids.emplace(layout.name(), vast::ids{}).first;
+  auto& ids = it->second;
+  VAST_ASSERT(first >= ids.size());
+  ids.append_bits(false, first - ids.size());
+  ids.append_bits(true, last - first);
+  if (is_new) {
+    // Insert new type.
+    for (size_t idx = 0; idx < layout.fields.size(); ++idx) {
+      auto& field = layout.fields[idx];
+      auto fqf = to_fully_qualified(layout.name(), field);
+      auto j = indexers_.find(fqf);
+      if (j == indexers_.end()) {
+        // Spawn a new indexer.
+        auto k = indexers_.emplace(fqf, wrapped_indexer{});
+        auto& ip = k.first->second;
+        auto pos = std::distance(indexers_.begin(), k.first);
+        ip.indexer = state().make_indexer(column_file(fqf), fqf.type, id(),
+                                          &measurements_[pos]);
+        state_->active_partition_indexers++;
+        ip.slot
+          = out().parent()->add_unchecked_outbound_path<table_slice_column>(
+            ip.indexer);
+        ip.outbound = out().paths().rbegin()->second.get();
```

Comment:
I didn't look up the type without `.get()`, but is it maybe possible to use the smart pointer in the indexer state instead of a raw pointer?

Reply:
It is a `unique_ptr<outbout_path>`, so copy is a no.